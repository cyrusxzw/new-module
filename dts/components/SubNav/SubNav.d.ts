export function getLinkItems(links: any, theme: any): any;
export const SubNav: React.ForwardRefExoticComponent<React.RefAttributes<any>>;
import React from 'react';
