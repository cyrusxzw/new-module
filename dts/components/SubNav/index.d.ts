import { SubNav } from './SubNav';
export { SubNav };
