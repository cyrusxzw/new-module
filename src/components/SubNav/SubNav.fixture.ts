import { HYPERLINK_STYLE_TYPES } from '~/constants';

const { EXTERNAL_NO_ICON_TEXT_LINK } = HYPERLINK_STYLE_TYPES;

const SubNavFixture = {
  heading: 'Phasellus aliquet sem',
  id: 'sub-nav',
  isSelect: false,
  links: [
    {
      id: 'Morbi-id-nulla',
      style: EXTERNAL_NO_ICON_TEXT_LINK,
      children: 'Morbi id nulla',
      url: '/',
    },
    {
      id: 'Nullam-ultrices',
      style: EXTERNAL_NO_ICON_TEXT_LINK,
      children: 'Nullam ultrices',
      url: '/',
    },
  ],
  theme: 'dark',
};

export { SubNavFixture };
