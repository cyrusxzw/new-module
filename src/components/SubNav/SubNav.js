import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import { useWindowHasResized } from '~/customHooks';
import { isViewport } from '~/utils/viewport';
import { List } from '~/components/List';
import { Select } from '~/components/Select';
import { Heading } from '~/components/Heading';
import { Hyperlink } from '~/components/Hyperlink';
import styles from './SubNav.module.css';

const SubNav = forwardRef(
  (
    { className, heading, headingClassName, id, isSelect, links, theme },
    ref,
  ) => {
    const currentTheme = useThemeContext(theme, 'dark');

    useWindowHasResized();

    const classSet = cx(
      styles.base,
      styles[currentTheme],
      { [styles.select]: isSelect },
      className,
    );

    const onChange = event => {
      window.location.href = event.target.value;
    };

    return (
      <nav className={classSet} ref={ref}>
        {heading && (
          <Heading
            className={cx(styles.heading, headingClassName)}
            level="3"
            noMargin={true}
            size="small"
            theme={currentTheme}
          >
            {heading}
          </Heading>
        )}
        {isSelect && isViewport('xs to md only') ? (
          <Select
            isBlock={isSelect}
            name={id}
            onChange={onChange}
            options={links.map(({ children, id, url }) => ({
              id,
              label: children,
              value: url,
            }))}
            theme={currentTheme}
          />
        ) : (
          <List
            items={getLinkItems(links, currentTheme)}
            listItemClassName={styles.item}
            theme={currentTheme}
          />
        )}
      </nav>
    );
  },
);

export function getLinkItems(links, theme) {
  return links.map(
    ({ children, id, hasTargetInNewWindow, style, url, onClick }) => ({
      content: (
        <Hyperlink
          className={styles.link}
          hasTargetInNewWindow={hasTargetInNewWindow}
          onClick={onClick}
          style={style}
          theme={theme}
          url={url}
        >
          {children}
        </Hyperlink>
      ),
      id,
    }),
  );
}

SubNav.propTypes = {
  className: PropTypes.string,
  id: PropTypes.string,
  isSelect: PropTypes.bool,
  links: PropTypes.arrayOf(PropTypes.object).isRequired,
  heading: PropTypes.string,
  headingClassName: PropTypes.string,
  theme: PropTypes.oneOf(['dark', 'light']),
};

SubNav.defaultProps = {
  className: undefined,
  id: undefined,
  isSelect: false,
  links: undefined,
  heading: undefined,
  headingClassName: undefined,
  theme: undefined,
};

export { SubNav };
