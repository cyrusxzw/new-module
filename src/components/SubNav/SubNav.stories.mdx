import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { StorybookGridOverlay } from '~.storybook/components';
import { SubNav } from './SubNav';
import { SubNavFixture } from './SubNav.fixture';

<Meta title="Components/SubNav" component={SubNav} />

# SubNav

<Canvas>
  <Story name="Base component">
    {(() => {
      const isSelect = knobs.boolean('isSelect', false, 'Presentation');
      const theme = knobs.select(
        'theme',
        ['dark', 'light'],
        'dark',
        'Presentation',
      );
      const implementation = {
        gridOverlay: {
          isActive: knobs.boolean('Show grid overlay', false, 'Implementation'),
          hasInvertedColours: knobs.boolean(
            'Invert grid overlay colours',
            false,
            'Implementation',
          ),
        },
      };
      return (
        <>
          <SubNav
            heading={SubNavFixture.heading}
            id={SubNavFixture.id}
            isSelect={isSelect}
            links={SubNavFixture.links}
            theme={theme}
          />
          <StorybookGridOverlay
            isActive={implementation.gridOverlay.isActive}
            hasInvertedColours={implementation.gridOverlay.hasInvertedColours}
          />
        </>
      );
    })()}
  </Story>
</Canvas>

## API

<ArgsTable of={SubNav} />
