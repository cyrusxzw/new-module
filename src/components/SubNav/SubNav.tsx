import React, { forwardRef } from 'react';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import { useWindowHasResized } from '~/customHooks';
import { isViewport } from '~/utils/viewport';
import { List } from '~/components/List';
import { Select } from '~/components/Select/index';
import { Heading } from '~/components/Heading/index';
import { getLinkItems, getSelectOptions, handleOnChange } from './SubNav.utils';
import type { SubNavProps } from './SubNav.types';
import styles from './SubNav.module.css';

const SubNav = forwardRef<HTMLElement, SubNavProps>(function SubNavRef(
  {
    className,
    heading,
    headingClassName,
    id,
    isSelect,
    links,
    theme,
    onSelectCallback,
  },
  ref,
) {
  const currentTheme = useThemeContext(theme, 'dark');

  useWindowHasResized();

  const classSet = cx(
    styles.base,
    styles[currentTheme],
    { [styles.select]: isSelect },
    className,
  );

  return (
    <nav className={classSet} ref={ref}>
      {heading && (
        <Heading
          className={cx(styles.heading, headingClassName)}
          level="3"
          size="small"
          theme={currentTheme}
        >
          {heading}
        </Heading>
      )}
      {isSelect && isViewport('xs to md only') ? (
        <Select
          isBlock={isSelect}
          name={id}
          onChange={(event) => {
            onSelectCallback?.(event);
            handleOnChange(event);
          }}
          options={getSelectOptions(links)}
          theme={currentTheme}
        />
      ) : (
        <List
          items={getLinkItems(links, currentTheme)}
          listItemClassName={styles.item}
          theme={currentTheme}
        />
      )}
    </nav>
  );
});

export { SubNav };
