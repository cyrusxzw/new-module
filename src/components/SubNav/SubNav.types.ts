import { MouseEvent, ReactNode } from 'react';
import type { Themes } from '~/types';
import type { LinkStyle } from '~/components/Hyperlink/Hyperlink.types';

type SubNavLink = {
  children: string /** @todo change this prop api to content or text  */;
  id?: string;
  hasTargetInNewWindow?: boolean;
  style?: LinkStyle;
  url: string;
  onClick?: (event: MouseEvent) => void;
};

type GetLinkItems = (
  links: SubNavLink[],
  theme: Themes,
) => { content: ReactNode; id: string }[];

type GetSelectOption = (
  links: SubNavLink[],
) => { id: string; label: string; value: string }[];

type SubNavProps = {
  className?: string;
  id: string;
  isSelect?: boolean;
  links: SubNavLink[];
  heading?: string;
  headingClassName?: string;
  theme?: Themes;
};

export type { SubNavProps, SubNavLink, GetLinkItems, GetSelectOption };
