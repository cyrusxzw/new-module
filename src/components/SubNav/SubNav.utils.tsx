import React from 'react';
import { Hyperlink } from '~/components/Hyperlink';
import type { ChangeEvent } from 'react';
import type { GetLinkItems, GetSelectOption } from './SubNav.types';
import styles from './SubNav.module.css';

const handleOnChange = (event: ChangeEvent<HTMLSelectElement>) => {
  window.location.href = event.target.value;
};

const getLinkItems: GetLinkItems = (links, theme) =>
  links.map(({ children, id, hasTargetInNewWindow, style, url, onClick }) => ({
    content: (
      <Hyperlink
        className={styles.link}
        hasTargetInNewWindow={hasTargetInNewWindow}
        onClick={onClick}
        style={style}
        theme={theme}
        url={url}
      >
        {children}
      </Hyperlink>
    ),
    id,
  }));

const getSelectOptions: GetSelectOption = (links) =>
  links.map(({ children, id, url }) => ({
    id,
    label: children,
    value: url,
  }));

export { getLinkItems, getSelectOptions, handleOnChange };
