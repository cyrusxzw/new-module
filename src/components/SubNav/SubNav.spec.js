import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';
import { SubNav } from './SubNav';
import { SubNavFixture } from './SubNav.fixture';

describe('<SubNav />', () => {
  it('should be defined', () => {
    expect(SubNav).toBeDefined();
  });

  it('should render base component correctly', () => {
    const { container } = render(
      <SubNav
        heading={SubNavFixture.heading}
        id={SubNavFixture.id}
        links={SubNavFixture.links}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render with corrent number of items and heading if passed', () => {
    render(
      <SubNav
        heading="test heading"
        id={SubNavFixture.id}
        links={SubNavFixture.links}
      />,
    );

    const heading = screen.getByRole('heading', { label: /heading/i });
    const items = screen.getAllByRole('listitem');

    expect(heading).toBeInTheDocument();
    expect(items).toHaveLength(SubNavFixture.links.length);
  });

  it('should render without heading if no `heading` prop is passed', () => {
    render(<SubNav links={SubNavFixture.links} />);

    const heading = screen.queryByRole('heading', { label: /heading/i });

    expect(heading).not.toBeInTheDocument();
  });

  it('should be accesible as a list', async () => {
    const { container } = render(
      <SubNav id="test-nav" isSelect={true} links={SubNavFixture.links} />,
    );

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });
});

describe('<SubNav isSelect />', () => {
  beforeAll(() => {
    // Force matchMedia to evaluate to true
    window.matchMedia = () => ({
      matches: true,
      addListener: () => undefined,
      removeListener: () => undefined,
    });
  });

  it('should render base component correctly', () => {
    const { container } = render(
      <SubNav
        heading={SubNavFixture.heading}
        id={SubNavFixture.id}
        isSelect={true}
        links={SubNavFixture.links}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render a select if isSelect is passed', () => {
    render(
      <SubNav id="test-nav" isSelect={true} links={SubNavFixture.links} />,
    );

    const nav = screen.getByRole('navigation');
    const select = screen.getByRole('combobox');
    const options = screen.getAllByRole('option');

    expect(nav).toBeInTheDocument();
    expect(select).toBeInTheDocument();
    expect(options).toHaveLength(SubNavFixture.links.length); // 2
  });

  it('should change window location on nav select change', () => {
    const mockResponse = jest.fn();

    render(
      <SubNav id="test-nav" isSelect={true} links={SubNavFixture.links} />,
    );

    Object.defineProperty(window, 'location', {
      value: {
        hash: {
          endsWith: mockResponse,
          includes: mockResponse,
        },
        assign: mockResponse,
      },
      writable: true,
    });

    const select = screen.getByRole('combobox');

    userEvent.selectOptions(select, 'https://aesop.com/');

    expect(window.location.href).toBe('https://aesop.com/');
  });
});

describe('<SubNav isSelect /> Accessibility', () => {
  beforeAll(() => {
    // https://github.com/nickcolley/jest-axe/issues/147
    window.getComputedStyle = () => undefined;

    // Force matchMedia to evaluate to true
    window.matchMedia = () => ({
      matches: true,
      addListener: () => undefined,
      removeListener: () => undefined,
    });
  });

  it('should be accesible as a nav select', async () => {
    const { container } = render(
      <SubNav id="test-nav" isSelect={true} links={SubNavFixture.links} />,
    );

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });
});
