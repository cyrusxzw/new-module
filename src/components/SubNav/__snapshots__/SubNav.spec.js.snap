// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubNav /> should render base component correctly 1`] = `
<div>
  <nav
    class="base dark"
  >
    <h3
      class="base small dark heading"
    >
      Phasellus aliquet sem
    </h3>
    <ul
      class="base dark"
    >
      <li
        class="item item"
      >
        <a
          class="base left dark link"
          href="https://aesop.com/"
          target="_self"
        >
          Morbi id nulla
        </a>
      </li>
      <li
        class="item item"
      >
        <a
          class="base left dark link"
          href="https://aesop.com/"
          target="_self"
        >
          Nullam ultrices
        </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`<SubNav isSelect /> should render base component correctly 1`] = `
<div>
  <nav
    class="base dark select"
  >
    <h3
      class="base small dark heading"
    >
      Phasellus aliquet sem
    </h3>
    <div
      class="base dark"
    >
      <label
        class="label"
        for="sub-nav"
      />
      <select
        aria-describedby="sub-nav-error-message"
        aria-invalid="false"
        aria-label="sub-nav"
        class="input isBlock"
        id="sub-nav"
        name="sub-nav"
      >
        <option
          value="https://aesop.com/"
        >
          Morbi id nulla
        </option>
        <option
          value="https://aesop.com/"
        >
          Nullam ultrices
        </option>
      </select>
      <svg
        aria-labelledby="chevron-1"
        class="base chevron dark icon"
        data-testid="data-testid-Icon"
        focusable="false"
        height="15"
        role="img"
        style="height: 15px; width: 15px;"
        tabindex="-1"
        viewBox="0 0 50 50"
        width="15"
      >
        <title
          id="chevron-1"
        >
          chevron
        </title>
        <g>
          <polygon
            class="chevron"
            points="25,31.3 4.2,10.5 0.1,14.6 25,39.5 50,14.6 45.9,10.5"
          />
        </g>
      </svg>
      
    </div>
  </nav>
</div>
`;
